### ISSUE TITLE
Will blue and black penmark be able to detect?

### ISSUE THREAD

**Issue by Cirno-tensai9** [choosehappy/HistoQC#250]
I am using histoQC, which is quite good, but our new sample contains plenty of black and blue pen marks, I tried to add image and mask to the data/pen/ and add to the config file. 

But it always show errors when adding >=3 pen mark types

2023-04-27 17:34:04,742 - INFO - sample/TCGA-A1-A0SF-01Z-00-DX1.7F252D89-EA78-419F-A969-1B7313D77499.sv                                                          s <class 'ValueError'> Inappropriate argument value (of correct type). cannot reshape array of size 162                                                          05760 into shape (2648,2040,2)
<code object byExampleWithFeatures at 0x560dd5459720, file "~/HistoQC/h                                                          istoqc/ClassificationModule.py", line 150>

**Comment by choosehappy** [comment]
Should be able to work, can you send us your data to test? the config file + masks?

**Comment by CielAl** [comment]
> Should be able to work, can you send us your data to test? the config file + masks?

3972 x 2040 x 2 == 16205760, and the shape provided by the OP is (2648 x 2040 x 2)

**Comment by choosehappy** [comment]
hmmm..its not clear me why there would be different sizes?

**Comment by Cirno-tensai9** [comment]
The files are in this shapes and all of them can work individually:
>>> black_pen.shape
(2231, 644, 3)
>>> black_mask.shape
(2231, 644, 3)
>>> blue_pen.shape
(587, 1000, 3)
>>> blue_mask.shape
(587, 1000, 3)
>>> red_pen.shape
(1000, 918, 3)
>>> red_mask.shape
(1000, 918, 3)
>>> green_pen.shape
(853, 1000, 3)
>>> green_mask.shape
(853, 1000, 3)

I added a blue pen mark as we have a lot sampels with blue pen mark and all the other colours are as provided.

Here is one of the config files that will cause reshape error (blue+red), green+red+black will also cause similar problem
[ClassificationModule.byExampleWithFeatures:pen_markings]
name: pen_markings
threshold: .8
examples: ./pen/1k_version/pen_blue.png:./pen/1k_version/pen_blue_mask.png
          ./pen/1k_version/pen_red.png:./pen/1k_version/pen_red_mask.png

area_threshold: 100
features:  frangi
           laplace
           rgb
           #lbp
           #gabor
           #median
           #gaussian

laplace_ksize: 3

frangi_scale_range: (1,10)
frangi_scale_step: 2
frangi_beta1: .5
frangi_beta2: 15
frangi_black_ridges: True

gabor_theta: 4
gabor_sigma: (1,3)
gabor_frequency: (0.05, 0.25)

lbp_radius: 3
lbp_points: 24
lbp_method: default

median_disk_size: 3

#gaussian_sigma: 1
#gaussian_multichan: False



[config_ML_test.txt](https://github.com/choosehappy/HistoQC/files/11396307/config_ML_test.txt)

**Comment by Cirno-tensai9** [comment]
I found the problem:
Differences occurs in step cal = clf.predict_proba(feats.reshape(-1, feats.shape[2])) 
in file histoqc/ClassificationModule.py

I used the sample : TCGA-A1-A0SF-01Z-00-DX1.7F252D89-EA78-419F-A969-1B7313D77499.svs
In some combinations (red+green): 
The cal shape is (5401920, 2)

But when mixed with (red+green+black):
The cal shape is (5401920,3)

But I have no idea where this different was from

**Comment by Cirno-tensai9** [comment]
Dear both,

I solved the problem, by cutting the samples and masks into smaller size.

Looks like the algorithm only detect the input samples and stacked the features together, and when the size is too large, it will think the training will need another dimension, but if cut it into smaller ones, it will run well and mine can currently detect 4 pen mark colours.

Best wishes,
Mengyuan

**Comment by choosehappy** [comment]
Great! Thanks for providing an update, glad it worked out!

**Comment by GuoJisen** [comment]
> Great! Thanks for providing an update, glad it worked out!

I also encountered this problem. Later, I compared my own mask with the mask that comes with the project, and found the problem.
The mask that comes with the project is a 1-bit color map, while my own mask is an 8-bit color map. If the input mask is a mixture of the two, it will cause this problem. My method is to convert the mask of the project to an 8-bit color map, and this problem is solved.
I hope it will be helpful to colleagues who encounter this problem in the future.
